v2.go

```diff
@@ -83,63 +88,67 @@ func parseVersion2(reader *bufio.Reader) (header *Header, err error) {
 	}
 	header.TransportProtocol = AddressFamilyAndProtocol(b14)
 	//  UNSPEC is only supported when LOCAL is set.
-	if header.TransportProtocol == UNSPEC {
-		if header.Command != LOCAL {
-			return nil, ErrUnsupportedAddressFamilyAndProtocol
-		}
-		// Ignore everything else.
-		return header, nil
+	if header.TransportProtocol.IsUnspec() && (header.Command != LOCAL || header.TransportProtocol != UNSPEC) {
+		return nil, ErrUnsupportedAddressFamilyAndProtocol
 	}
 
 	// Make sure there are bytes available as specified in length
 	var length uint16
 	if err := binary.Read(io.LimitReader(reader, 2), binary.BigEndian, &length); err != nil {
 		return nil, ErrCantReadLength
 	}
-	if !header.validateLength(length) {
+
+	// There's no need to validate for UNSPEC since length always >= 0
+	if header.TransportProtocol != UNSPEC && !header.validateLength(length) {
 		return nil, ErrInvalidLength
 	}
 
+	if length == 0 {
+		return header, nil
+	}
+
 	if _, err := reader.Peek(int(length)); err != nil {
 		return nil, ErrInvalidLength
 	}
 
 	// Length-limited reader for payload section
 	payloadReader := io.LimitReader(reader, int64(length)).(*io.LimitedReader)
 
-	// Read addresses and ports
-	if header.TransportProtocol.IsIPv4() {
-		var addr _addr4
-		if err := binary.Read(payloadReader, binary.BigEndian, &addr); err != nil {
-			return nil, ErrInvalidAddress
-		}
-		header.SourceAddr = newIPAddr(header.TransportProtocol, addr.Src[:], addr.SrcPort)
-		header.DestinationAddr = newIPAddr(header.TransportProtocol, addr.Dst[:], addr.DstPort)
-	} else if header.TransportProtocol.IsIPv6() {
-		var addr _addr6
-		if err := binary.Read(payloadReader, binary.BigEndian, &addr); err != nil {
-			return nil, ErrInvalidAddress
-		}
-		header.SourceAddr = newIPAddr(header.TransportProtocol, addr.Src[:], addr.SrcPort)
-		header.DestinationAddr = newIPAddr(header.TransportProtocol, addr.Dst[:], addr.DstPort)
-	} else if header.TransportProtocol.IsUnix() {
-		var addr _addrUnix
-		if err := binary.Read(payloadReader, binary.BigEndian, &addr); err != nil {
-			return nil, ErrInvalidAddress
-		}
+	// Read addresses and ports for protocols other than UNSPEC
+	if header.TransportProtocol != UNSPEC {
```

Comment:
Please, add a comment stating if UNSPEC, this code, as the receiver, ignore address information and will skip straight to reading TLVs because length is greater than zero and there should be TLVs.